grails-app/services/org/pih/warehouse/product/ProductService.groovy

```diff
@@ -1025,67 +1073,46 @@ class ProductService {
         return getPopularTags(0)
     }
 
-
     /**
-     * Add a list of tags to each of the given products.
-     *
-     * @param products
-     * @param tags
-     * @return
+     * Find or create a list of product tags with the given names as a map keyed on tag name. Nulls, duplicates,
+     * and whitespace will be ignored.
      */
-    def addTagsToProducts(products, tags) {
-        log.info "Add tags ${tags} to products ${products}"
-        products.each { product ->
-            addTagsToProduct(product, tags)
-        }
+    private Map<String, Tag> getOrCreateTagsAsMap(List<String> tagNames) {
+        return getOrCreateTags(tagNames).collectEntries{ [ it.tag, it ] }
     }
 
     /**
-     * Add the list of tags to the given product.
-     *
-     * @param product
-     * @param tags
+     * Find or create a list of product tags with the given names. Nulls, duplicates, and whitespace will be ignored.
      */
-    def addTagsToProduct(product, tags) {
-        log.info "Add tags ${tags} to product ${product}"
-        if (tags) {
-            tags.each { tagName ->
-                if (tagName) {
-                    addTagToProduct(product, tagName)
-                }
-            }
+    private List<Tag> getOrCreateTags(List<String> tagNames) {
+        List<String> tagNamesSanitized = removeBlanksAndDuplicates(tagNames)
+        if (!tagNamesSanitized) {
+            return Collections.emptyList()
+        }
+
+        List<Tag> tags = []
+        for (tagName in tagNamesSanitized) {
+            Tag tag = findOrCreateTag(tagName)
+            tags.add(tag)
         }
+        return tags
     }
 
     /**
-     * Add the given tag to the given product.
-     *
-     * @param product
-     * @param tagName
-     * @return
+     * Find or create a tag with the given name.
      */
-    def addTagToProduct(product, tagName) {
-        log.info "Add tags ${tagName} to product ${product}"
-
-        // Check if the product already has the given tag
-        def tag = product.tags.find { it.tag == tagName }
-
-        if (!tag) {
-            // Otherwise try to find an existing tag that matches the tag
-            tag = Tag.findByTag(tagName)
-            // Or create a brand new one
-            if (!tag) {
-                log.info "Tag ${tagName} does not exist so creating it"
-                tag = new Tag(tag: tagName)
-                tag.save(flush: true)
-            }
-            product.addToTags(tag)
-            product.save(flush: true)
-        } else {
-            log.info "Product ${product} already contains tag ${tag}"
+    private Tag findOrCreateTag(String tagName) {
+        Tag tag = Tag.findByTag(tagName)
+        if (tag) {
+            return tag
         }
-    }
 
+        tag = new Tag(tag: tagName)
+        if (!tag.save()) {
```

Comment:
Note that after this refactor, we don't need to flush the cache when saving tags anymore since we're directly returning and using them (whereas before we'd try to fetch from the db later and so we had to ensure the db session was flushed).